### DISCUSSION TITLE
how to set priors when I don't have any for Multiple Parameters are Regression Targets in HSSM

### DISCUSSION THREAD

**Discussion by XiaoyuZeng** [lnccbrown/HSSM#392] [Q&A] [answered]
Dear HSSM team,

I am trying to fit a DDM with four parameters being regressed with two factors and their interaction. This model was fit at both group and individual level. 
I was asked to add the prior for each of the intercept as well as regressor. I even found I have to set these priors for individual-level random slope too. 
I even noticed you set different priors for the intercept for different parameters.

information about the regressors:
condition and stage are both categorical variables
![image](https://github.com/lnccbrown/HSSM/assets/38243112/e05e31cc-495b-4dfc-ac30-ee5bb3d0353f)

**my question is, how should I set all of these priors of group-level and individual-level and for different parameters (v,z,a,d) when I actually don't have any?**

included is my current priors, correct me if I was wrong about this setting. 
also, I am looking forward for your suggestions to optimize this setting.

`mfull = hssm.HSSM(model="ddm",
                  data=df,
                 include=[
        {
            "name": "v",
            "prior": {
                "Intercept": {"name": "Uniform", "lower": -3.0, "upper": 3.0},
                "condition": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
                "stage": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
                "condition:stage": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
            },
            "formula": "v ~  condition + stage + condition * stage + (1+condition * stage|subj_idx)",
            "link": "identity",
        },
         {
            "name": "z",
            "prior": {
                "Intercept": {"name": "Uniform", "lower": 0, "upper": 1},
                "condition": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
                "stage": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
                 "condition:stage": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
            },
            "formula": "z ~  condition + stage + condition * stage + (1+condition * stage|subj_idx)",
            "link": "identity",
        },
        {
            "name": "a",
            "prior": {
                "Intercept": {"name": "Uniform", "lower": 0.5, "upper": 3},
                "condition": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
                "stage": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
                 "condition:stage": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
            },
            "formula": "a ~  condition + stage + condition * stage + (1+condition * stage|subj_idx)",
            "link": "identity",
        },
        {
            "name": "t",
            "prior": {
                "Intercept": {"name": "Uniform", "lower": 0.0015, "upper": 2},
                "condition": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
                "stage": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
                 "condition:stage": {"name": "Uniform", "lower": -1.0, "upper": 1.0},
            },
            "formula": "t ~  condition + stage + condition * stage + (1+condition * stage|subj_idx)",
            "link": "identity",
        },
    ],)`

**Comment by XiaoyuZeng** [comment]
Hi all,
I am also trying a version where I just ignored the priors, it seems that the fault prior settings are mostly normal. 
we will see how this version goes.


`Hierarchical Sequential Sampling Model
Model: ddm

Response variable: rt,response
Likelihood: analytical
Observations: 5215

Parameters:

v:
    Formula: v ~  1+ condition + stage + condition * stage + (1+condition * stage|subj_idx)
    Priors:
        v_Intercept ~ Normal(mu: 0.0, sigma: 4.4969000816345215)
        v_condition ~ Normal(mu: 0.0, sigma: 5.003200054168701)
        v_stage ~ Normal(mu: 0.0, sigma: 5.000199794769287)
        v_condition:stage ~ Normal(mu: 0.0, sigma: 5.8246002197265625)
        v_1|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 4.4969000816345215))
        v_condition|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 5.003200054168701))
        v_stage|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 5.000199794769287))
        v_condition:stage|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 5.8246002197265625))
    Link: identity
    Explicit bounds: (-inf, inf)
a:
    Formula: a ~ 1+ condition + stage + condition * stage + (1+condition * stage|subj_idx)
    Priors:
        a_Intercept ~ Normal(mu: 0.0, sigma: 1.0)
        a_condition ~ Normal(mu: 0.0, sigma: 1.0)
        a_stage ~ Normal(mu: 0.0, sigma: 1.0)
        a_condition:stage ~ Normal(mu: 0.0, sigma: 1.0)
        a_1|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        a_condition|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        a_stage|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        a_condition:stage|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
    Link: identity
    Explicit bounds: (0.0, inf)
z:
    Formula: z ~  1+condition + stage + condition * stage + (1+condition * stage|subj_idx)
    Priors:
        z_Intercept ~ Normal(mu: 0.0, sigma: 1.0)
        z_condition ~ Normal(mu: 0.0, sigma: 1.0)
        z_stage ~ Normal(mu: 0.0, sigma: 1.0)
        z_condition:stage ~ Normal(mu: 0.0, sigma: 1.0)
        z_1|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        z_condition|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        z_stage|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        z_condition:stage|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
    Link: identity
    Explicit bounds: (0.0, 1.0)
t:
    Formula: t ~ 1+ condition + stage + condition * stage + (1+condition * stage|subj_idx)
    Priors:
        t_Intercept ~ Normal(mu: 0.0, sigma: 1.0)
        t_condition ~ Normal(mu: 0.0, sigma: 1.0)
        t_stage ~ Normal(mu: 0.0, sigma: 1.0)
        t_condition:stage ~ Normal(mu: 0.0, sigma: 1.0)
        t_1|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        t_condition|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        t_stage|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
        t_condition:stage|subj_idx ~ Normal(mu: 0.0, sigma: HalfNormal(sigma: 1.0))
    Link: identity
    Explicit bounds: (0.0, inf)

Lapse probability: 0.05
Lapse distribution: Uniform(lower: 0.0, upper: 10.0)`

**Reply by XiaoyuZeng** [reply]
unfortunately, this version leads to  error: 

> SamplingError: Initial evaluation of model at starting point failed!
> Logp initial evaluation results:
> {'v_Intercept': -2.42, 'v_condition': -2.53, 'v_stage': -2.53, 'v_condition:stage': -2.68, 'v_1|subj_idx_sigma': -0.73, 'v_1|subj_idx_offset': -45.03, 'v_condition|subj_idx_sigma': -0.73, 'v_condition|subj_idx_offset': -45.03, 'v_stage|subj_idx_sigma': -0.73, 'v_stage|subj_idx_offset': -45.03, 'v_condition:stage|subj_idx_sigma': -0.73, 'v_condition:stage|subj_idx_offset': -45.03, 'a_Intercept': -0.92, 'a_condition': -0.92, 'a_stage': -0.92, 'a_condition:stage': -0.92, 'a_1|subj_idx_sigma': -0.73, 'a_1|subj_idx_offset': -45.03, 'a_condition|subj_idx_sigma': -0.73, 'a_condition|subj_idx_offset': -45.03, 'a_stage|subj_idx_sigma': -0.73, 'a_stage|subj_idx_offset': -45.03, 'a_condition:stage|subj_idx_sigma': -0.73, 'a_condition:stage|subj_idx_offset': -45.03, 'z_Intercept': -0.92, 'z_condition': -0.92, 'z_stage': -0.92, 'z_condition:stage': -0.92, 'z_1|subj_idx_sigma': -0.73, 'z_1|subj_idx_offset': -45.03, 'z_condition|subj_idx_sigma': -0.73, 'z_condition|subj_idx_offset': -45.03, 'z_stage|subj_idx_sigma': -0.73, 'z_stage|subj_idx_offset': -45.03, 'z_condition:stage|subj_idx_sigma': -0.73, 'z_condition:stage|subj_idx_offset': -45.03, 't_Intercept': -0.92, 't_condition': -0.92, 't_stage': -0.92, 't_condition:stage': -0.92, 't_1|subj_idx_sigma': -0.73, 't_1|subj_idx_offset': -45.03, 't_condition|subj_idx_sigma': -0.73, 't_condition|subj_idx_offset': -45.03, 't_stage|subj_idx_sigma': -0.73, 't_stage|subj_idx_offset': -45.03, 't_condition:stage|subj_idx_sigma': -0.73, 't_condition:stage|subj_idx_offset': -45.03, 'rt,response': inf}
> You can call `model.debug()` for more details.

do you think it's due to the complexity of the hierarchical regression model? do you have any idea to improve this?

**Comment by digicosmos86** [comment, marked as answer]
Hi @XiaoyuZeng,

Thank you for using the HSSM package! First, you can simplify the right hand side of all of your formulas to `1 + condition * stage + (1+condition * stage|subj_idx)`, because when specifying interactions, `a * b` = `a + b + a:b` where `a:b` is the interaction term.

We do have a set of default priors when priors are not provided. We have recently updated the package but these changes have not really made it to the officially released package yet. To incorporate these changes, you can run `pip install git+https://github.com/lnccbrown/HSSM.git` to install the dev version of HSSM. After that, you can add this argument `prior_settings="safe"` when you call `hssm.HSSM()`, and the default priors will be replaced by a `safe` version provided by us.

The default priors when you don't specify `safe` priors are set by `bambi`, which does not take into account of the different requirements of the parameters of DDM, which could cause sampling to fail.

Hope this helps. Let me know if this works :)

Thanks!
Paul

**Comment by XiaoyuZeng** [comment]
> Hi @XiaoyuZeng,
> 
> Thank you for using the HSSM package! First, you can simplify the right hand side of all of your formulas to `1 + condition * stage + (1+condition * stage|subj_idx)`, because when specifying interactions, `a * b` = `a + b + a:b` where `a:b` is the interaction term.
> 
> We do have a set of default priors when priors are not provided. We have recently updated the package but these changes have not really made it to the officially released package yet. To incorporate these changes, you can run `pip install git+https://github.com/lnccbrown/HSSM.git` to install the dev version of HSSM. After that, you can add this argument `prior_settings="safe"` when you call `hssm.HSSM()`, and the default priors will be replaced by a `safe` version provided by us.
> 
> The default priors when you don't specify `safe` priors are set by `bambi`, which does not take into account of the different requirements of the parameters of DDM, which could cause sampling to fail.
> 
> Hope this helps. Let me know if this works :)
> 
> Thanks! Paul

Many thanks for your help, Paul!

The argument `prior_settings="safe"` indeed is what I am looking for!

After a lot of struggles to install the latest hssm via git, I found this argument in the hssm 0.2.1 worked well. 

In addition, in this latest version of hssm, I seem to find a bug related to the random slope. I will submit another bug report about this.


Thanks again!

Best,
Xiaoyu
> Hierarchical Sequential Sampling Model
> Model: ddm
> 
> Response variable: rt,response
> Likelihood: analytical
> Observations: 5215
> 
> Parameters:
> 
> v:
>     Formula: v ~ 1 + condition * stage +(1|subj_idx)
>     Priors:
>         v_Intercept ~ Normal(mu: 0.0, sigma: 0.25)
>         v_condition ~ Normal(mu: 0.0, sigma: 0.25)
>         v_stage ~ Normal(mu: 0.0, sigma: 0.25)
>         v_condition:stage ~ Normal(mu: 0.0, sigma: 0.25)
>         v_1|subj_idx ~ Normal(mu: 0.0, sigma: Weibull(alpha: 1.5, beta: 0.30000001192092896))
>     Link: identity
>     Explicit bounds: (-inf, inf)
> 
> a:
>     Formula: a ~ 1 + condition * stage +(1|subj_idx))
>     Priors:
>         a_Intercept ~ Normal(mu: 0.0, sigma: 0.25)
>         a_condition ~ Normal(mu: 0.0, sigma: 0.25)
>         a_stage ~ Normal(mu: 0.0, sigma: 0.25)
>         a_condition:stage ~ Normal(mu: 0.0, sigma: 0.25)
>         a_1|subj_idx ~ Normal(mu: 0.0, sigma: Weibull(alpha: 1.5, beta: 0.30000001192092896))
>     Link: identity
>     Explicit bounds: (0.0, inf)
> 
> z:
>     Formula: z ~ 1 + condition * stage +(1|subj_idx))
>     Priors:
>         z_Intercept ~ Normal(mu: 0.0, sigma: 0.25)
>         z_condition ~ Normal(mu: 0.0, sigma: 0.25)
>         z_stage ~ Normal(mu: 0.0, sigma: 0.25)
>         z_condition:stage ~ Normal(mu: 0.0, sigma: 0.25)
>         z_1|subj_idx ~ Normal(mu: 0.0, sigma: Weibull(alpha: 1.5, beta: 0.30000001192092896))
>     Link: identity
>     Explicit bounds: (0.0, 1.0)
> 
> t:
>     Formula: t ~ 1 + condition * stage +(1|subj_idx))
>     Priors:
>         t_Intercept ~ Normal(mu: 0.0, sigma: 0.25)
>         t_condition ~ Normal(mu: 0.0, sigma: 0.25)
>         t_stage ~ Normal(mu: 0.0, sigma: 0.25)
>         t_condition:stage ~ Normal(mu: 0.0, sigma: 0.25)
>         t_1|subj_idx ~ Normal(mu: 0.0, sigma: Weibull(alpha: 1.5, beta: 0.30000001192092896))
>     Link: identity
>     Explicit bounds: (0.0, inf)
> 
> 
> Lapse probability: 0.05
> Lapse distribution: Uniform(lower: 0.0, upper: 10.0)
